doc/boot.conf.5
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "BOOT.CONF" 5 "" "2.8.7" "Funtoo Linux Core System"
.SH NAME
boot.conf \- Funtoo global boot loader configuration file
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
\fI/etc/boot.conf\fP
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
The data in \fI/etc/boot.conf\fP is used by \fIego\-boot(8)\fP to generate specific
configuration files for various boot loaders as necessary.
.sp
The benefit of \fI/etc/boot.conf\fP is that it provides a single location to
store all boot\-related information. It also provides a single, consistent file
format and feature set for configuring all boot loaders.
.sp
For MBR booting, GRUB, GRUB (legacy) and LILO boot loaders are supported. For
UEFI booting, GRUB is supported. When using GRUB, boot\-update will auto\-enable
UEFI boot configuration if it sees that your system has been booted in UEFI mode.
Otherwise, MBR (legacy) boot mode will be used.
.sp
\fIboot\-update(8)\fP utilizes \fI/etc/boot.conf\fP to provide a consistent process for
updating boot loader configuration, regardless of the actual boot loader used.
.sp
Here is a sample \fI/etc/boot.conf\fP configuration file:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
boot {
        generate grub
        # \(dqgenerate grub\-legacy\(dq if you are using grub\-legacy\-0.97

        # Below, you can specify the kernel image to boot by default,
        # or the default boot entry name

        default bzImage
        # default Windows 7
}

display {
        # gfxmode 1440x900
        # background /boot/background.jpg
}

\(dqAll my kernels\(dq {
        #
        # This entry use reasonable defaults to find all matching kernels.  An
        # \(dqAll my kernels \- <kernel name>\(dq boot entry will be generated for
        # *each* kernel found.
        #
        # type \(dqboot\-update \-\-showdefaults\(dq to see default settings.
}

\(dqFuntoo Linux console\(dq {
        kernel bzImage
}

\(dqFuntoo Linux with uvesafb\(dq {
        kernel bzImage\-RHEL5\-uvesafb[\-v]
        params += video=uvesafb:1440x900\-8,mtrr:2
}

\(dqXen\(dq {
        type xen
        xenkernel xen\-4.1.1.gz
        xenparams loglvl=all guest_loglvl=all
        kernel bzImage[\-v]
        initrd initramfs[\-v]
}

\(dqWindows 7\(dq {
        type win7
        params root=/dev/sda1
}

\(dqWindows XP\(dq {
        type winxp
        params root=/dev/sda4
}

.ft P
.fi
.UNINDENT
.UNINDENT
.SS Sections
.sp
Sections are specified by an alphanumeric name, followed by a space and a \fI{\fP\&.
A section ends when a single \fI}\fP appears on a line.
.sp
There are special \fIbuilt\-in\fP sections that are expected to be found by the
framework and used for configuation settings. You can not use the name of
a \fIbuilt\-in\fP section as the name of a boot entry.
Currently these are all \fIbuilt\-in\fP sections:
\fIboot\fP, \fIcolor\fP, \fIdefault\fP, \fIgrub\fP, \fIgrub\-legacy\fP, \fIlilo\fP, \fIserial\fP
.sp
In addition, other sections can be created. Any sections with non\-builtin names
are recognized as boot entry definitions. For example, the sections \fI\(dqFuntoo
Linux\(dq\fP and \fI\(dqFuntoo Linux genkernel\(dq\fP define boot entries. Due to
\fI/etc/boot.conf\fP\(aqs wildcard support, a single boot entry section in the
configuration file may generate multiple actual boot entries for the boot
loader, depending on how many matches are found. Wildcard support will be
explained later in this document.
.SS Default Section
.sp
In addition, there is a special section named \fIdefault\fP which is used to
specify default settings for the boot entry sections. \fIboot\-update\fP has several
reasonable default built\-in settings that can be overridden. For example, if
you leave out the \fIboot/timeout\fP setting, the boot menu timeout will default to
5 seconds. To see all built\-in settings, type \fIboot\-update \-\-showdefaults\fP\&.
.SS Linux Boot Entries
.sp
There are four critical parameters that are used in boot entry and \fIdefault\fP
sections \-\- \fItype\fP, \fIkernel\fP, \fIinitrd\fP and \fIparams\fP\&. \fItype\fP defaults
to \(dqlinux\(dq and informs \fIboot\-update\fP that we are specifying a Linux boot
entry.  It can be set to other values to tell \fIboot\-update\fP that we are
specifying a Microsoft Windows 7 boot entry, for example.
.SS Linux Kernel Wildcards
.sp
The \fIkernel\fP variable specifies one or more kernels, using exact kernel file
names or wildcards. Again, note that it is possible for one boot entry in
\fI/etc/boot.conf\fP to generate \fImultiple\fP boot entries for your boot loader if
wildcards are used or multiple kernels are listed \-\- one boot entry will be
generated for each matching kernel found.
.sp
So, for example, the following
\fI/etc/boot.conf\fP could generate two boot entries named \(dqFuntoo Linux \-
bzImage\(dq and \(dqFuntoo Linux \- bzImage\-new\(dq:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqFuntoo Linux\(dq {
        kernel bzImage bzImage\-new
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The \fI[\-v]\fP wildcard, which can only appear once per filename, can be used at
the end of a kernel image name to match the base name, or any combination of
the base name, plus a hypen and any additional text:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqFuntoo Linux\(dq {
        kernel bzImage[\-v]
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Above, \fIbzImage[\-v]\fP will match \fIbzImage\fP as well as \fIbzImage\-*\fP\&. Note that
any files ending in \fI\&.sig\fP will now be ignored.
.sp
In addition, \fIboot.conf\fP now supports the inclusion of arbitrary glob wildcards
within brackets, which work similarly to \fI[\-v]\fP, above:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqFuntoo Linux\(dq {
        kernel bzImage[\-2.6*]
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The above wildcard will match \(dqbzImage\(dq, \(dqbzImage\-2.6.18\(dq, and \(dqbzImage\-2.6.24\(dq.
.sp
You can also now exclude certain patterns by providing additional patterns
prefixed by \fI\-\fP, after the initial match(es):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqFuntoo Linux\(dq {
        kernel bzImage[\-2.6*] \-bzImage\-2.6.8 \-bzImage\-3*
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Remember that wildcards are optional. If you don\(aqt want to deal with them, you
can just provide the name of a kernel image.
.SS initrd/initramfs
.sp
The \fIinitrd\fP variable specifies one or more initrds or initramfs images, like
this:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqFuntoo Linux\(dq {
        kernel bzImage
        initrd initramfs.igz
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fIinitrd\fP also allows the use of the \fI[\-v]\fP wildcard to allow you to create
matching pairs of kernels and initrds on disk that boot\-update will associate
with one another automatically by suffix. Here\(aqs how it works \-\- assume you have
the following boot entry:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqFuntoo Linux\(dq {
        kernel bzImage[\-v]
        initrd initramfs[\-v]
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The \fI/etc/boot.conf\fP entry above will look for all kernels matching \fIbzImage\fP
and \fIbzImage\-*\fP and generate a boot entry for each one. For the boot entry for
\fIbzImage\fP, the \fIinitramfs[\-v]\fP wildcard will pull in the initramfs \fIinitramfs\fP
if it exists \-\- if not, it will be skipped. For the boot entry for
\fIbzImage\-2.6.24\fP, the initramfs \fIinitramfs\-2.6.24\fP will be used if it exists.
.sp
If you are using the enhanced glob wildcard functionality in your \fIkernel\fP
option (such as \fIbzImage[\-2.6\fP]*, above), then remember that you should still
use \fI[\-v]\fP in your \fIinitrd\fP option. \fI[\-v]\fP is the only pattern that is supported
for initrds.
.SS Multiple initrds
.sp
Since Linux allows multiple initramfs images to be loaded at boot time, you can
specify more than one initrd in a boot entry, and the specified initrds will be
loaded in succession abt boot time. Note that this is different from the
\fIkernel\fP option \- where multiple matches will generate multiple boot entries,
since you can only load one kernel at boot. Here\(aqs an example:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqFuntoo Linux\(dq {
        kernel bzImage
        initrd initramfs\-1.igz initramfs\-2.igz
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
In the above example, a single boot entry will be generated, which will load
\fIinitramfs\-1.igz\fP and \fIinitramfs\-2.igz\fP as the primary and secondary initramfs
respectively, and then boot the kernel \fIbzImage\fP\&.
.sp
Note that the \fI+=\fP operator can be used to either extend the default initramfs
setting or to specify multiple initramfs images over multiple lines. Here\(aqs
a boot entry that is equivalent to the previous example:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqFuntoo Linux\(dq {
        kernel bzImage
        # load initramfs\-1.igz:
        initrd initramfs\-1.igz
        # also load initramfs\-2.igz:
        initrd += initramfs\-2.igz
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
And in the following example, the initial \fI+=\fP tells coreboot to append
\fIinitramfs\-1.igz\fP to the default initramfs list:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqFuntoo Linux\(dq {
        kernel bzImage
        # load our default initramfs image(s), plus this one:
        initrd += initramfs\-1.igz
}
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Parameters
.sp
The \fIparams\fP variable specifies kernel parameters used to boot the kernel.
Typical kernel parameters, such as \fIinit=/bin/bash\fP, \fIroot=/dev/sda3\fP or others
can be specified as necessary. Here\(aqs a sophisticated example from Andreas
Matuschek that was posted on the funtoo\-dev mailing list:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqFuntoo Linux On Ice\(dq {
        params root=/dev/sda2
        params += rootfstype=jfs
        params += usbcore.autosuspend=1
        params += acpi_sleep=s3_bios,s3_mode
        params += hpet=force
        params += video=radeonfb:ywrap,mtrr:1,1024x768\-32@60
        params += quiet
        params += splash=silent,kdgraphics,theme:natural_gentoo
        params += CONSOLE=/dev/tty1
        kernel vmlinuz[\-v]
        initrd ramfs
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
As you can see, when  \fI+=\fP is used as the first argument for \fIparams\fP, the
default setting can be \fIextended\fP with additional parameters. If the first
\fIparams root=/dev/sda2\fP line was instead written as \fIparams += root=/dev/sda2\fP,
then all the parameters specified in this boot entry would \fIextend\fP the default
params settings. But in this case, Andreas specified the first \fIparams\fP
parameter in this boot entry without a \fI+=\fP, so his settings replace the
default settings.
.sp
Note that when serial console support is enabled, the appropriate parameters
for serial support (from the serial section) will be added to params.
.SS Special Parameters
.INDENT 0.0
.TP
.B \fB+=\fP
When \fI+=\fP is specified at the beginning of the first \fIparams\fP, \fIinitrd\fP or
\fIkernel\fP definition in a section, then the arguments after the \fI+=\fP will be
added to the default settings defined in \fIdefault\fP (type \fIboot\-update
\-\-showdefaults\fP to see default settings.)  In addition, multiple \fIparams\fP,
\fIinitrd\fP or \fIkernel\fP lines can appear in a section, as long as the successive
lines begin with \fI+=\fP\&. This allows these values to be defined over multiple
lines.
.TP
.B \fBroot=auto\fP
When \fIroot=auto\fP is evaluated, the framework will look at \fI/etc/fstab\fP to
determine the root filesystem device. Then \fIroot=auto\fP will changed to
reflect this, so the actual parameter passed to the kernel will be something
like \fIroot=/dev/sda3\fP .
.TP
.B \fBrootfstype=auto\fP
In a similar fashion to \fIroot=auto\fP, \fIrootfstype=auto\fP will be
replaced with something like \fIrootfstype=ext4\fP, with the filesystem type
determined by the setting in \fI/etc/fstab\fP\&.
.TP
.B \fBreal_root=auto\fP
This special parameter is useful when using \fIgenkernel\fP initrds that expect a
\fIreal_root\fP parameter. When specified, any \fIroot=\fP options already specified
(including \fIroot=auto\fP) will be removed from \fIparams\fP, and \fIreal_root\fP will
be set to the root filesystem based on \fI/etc/fstab\fP, so you\(aqll end up with a
setting such as \fIreal_root=/dev/sda3\fP\&.
.UNINDENT
.SH LINUX DISTRIBUTIONS ON SEPARATE PARTITIONS
.sp
\fIboot\-update\fP supports creating boot entries for Linux distributions installed
on separate partitions. In order for \fIboot\-update\fP to find the kernels and initrds
located on other partitions you have to create a mount point for the partition.
After creating a mount point you must specify the absolute path to the kernels
using the scan variable:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqDebian Sid\(dq {
        scan /mnt/debian/boot
        kernel vmlinuz[\-v]
        initrd initrd.img
        params root=/dev/sdb1
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Note that you must also set \fIparams root=/dev/<root>\fP to the correct root
partition in order to override the default \fIroot=auto\fP setting. At this time
\fIboot\-update\fP does not support auto detecting for other Linux systems.
.sp
If you would like boot\-update to auto mount the partition whenever it is ran, you
must create an entry for mounting it in \fI/etc/fstab\fP\&. Otherwise you will need to
mount the partition before running \fIboot\-update\fP\&.
.SH ALTERNATE OS LOADING
.sp
Boot entries can be created for alternate operating systems using the following
approach:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqWindows 7\(dq {
        type win7
        params root=/dev/sda6
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The \fItype\fP variable should be set to one of the operating system names that
\fIboot\-update\fP recognizes (case\-insensitive,) which are:
.INDENT 0.0
.IP \(bu 2
linux (default)
.IP \(bu 2
dos
.IP \(bu 2
msdos
.IP \(bu 2
Windows 2000
.IP \(bu 2
win2000
.IP \(bu 2
Windows XP
.IP \(bu 2
winxp
.IP \(bu 2
Windows Vista
.IP \(bu 2
vista
.IP \(bu 2
Windows 7
.IP \(bu 2
win7
.IP \(bu 2
Windows 8
.IP \(bu 2
win8
.IP \(bu 2
Windows 10
.IP \(bu 2
win10
.IP \(bu 2
Haiku
.IP \(bu 2
Haiku OS
.UNINDENT
.sp
For non\-Linux operating systems, the \fIparams\fP variable is used to specify the
root partition for chain loading. For consistency with Linux boot entries, the
syntax used is \fIroot=device\fP\&.
.sp
Users can manually set the chainloader option if they require a special value
other than the normal default set by boot.conf:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqWindows 10\(dq {
    type win10
    params root=/dev/sda6
    params += chainloader=+4
}
.ft P
.fi
.UNINDENT
.UNINDENT
.SH BOOT SECTION
.SS \fIboot :: autopick\fP
.sp
Specifies the strategy that is used to select the default kernel when no
explicit default is selected via the \fIboot :: default\fP option below.
This option defaults to \fImtime\fP, which means that the matching kernel with the most
recent modification time will be selected as a default kernel for booting.
.sp
Other possibilities for this option are \fIlast\-booted\fP, which will automatically
update the default kernel boot selection to match the last kernel booted. Note
that this option requires running \fIego boot success\fP after the system has
successfully booted, typically from \fI/etc/local.d\fP\&.
.SS \fIboot :: generate\fP
.sp
Specifies the boot loader that \fIboot\-update\fP should generate a configuration
files for. This setting should be a single string, set to one of \fIgrub\fP,
\fIgrub\-legacy\fP or \fIlilo\fP\&. Defaults to \fIgrub\fP\&.
.SS \fIboot :: timeout\fP
.sp
Specifies the boot loader timeout, in seconds. Defaults to \fI5\fP\&.
.SS \fIboot :: default\fP
.sp
Use this setting to specify the boot entry to boot by default. There are two
ways to use this setting.
.sp
The first way is to specify the filename of the kernel to boot by default. This
setting should contain no path information, just the kernel image name.  This
is the default mechanism, due to the setting of \fIbzImage\fP\&.
.sp
Alternatively, you can also specify the literal name of the boot entry you want
to boot. This is handy if you want to boot a non\-Linux operating system by
default. If you had the following boot entry:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqMy Windows 7\(dq {
        type win7
        params root=/dev/sda6
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\&...then, you could boot this entry by default with the following boot section:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
boot {
        generate grub
        default My Windows 7
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This is also a handy mechanism if you want to boot the most recently created
kernel by default. The kernel with the most recent modification time will be
selected. To do this, specify the name of the boot entry rather than
the kernel image name:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
boot {
        default \(dqFuntoo Linux\(dq
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If multiple \(dqFuntoo Linux\(dq boot entries are created, the one that has the most
recently created kernel (by file modification time) will be booted by default.
.sp
Also note that if no \fBdefault\fP setting is specified, or no match is found,
boot\-update will select the most recent kernel by file modification time and
set it as the default selection for booting.
.sp
Note that double\-quotes are optional both in section names and in the
\fIboot/default\fP value.
.SS \fIboot :: bootdev\fP
.sp
Specifies which device or partition to install the bootloader to. This is
currently only used for lilo and is the equivalent of setting \(dqboot = <bootdev>\(dq
in \fB/etc/lilo.conf\fP\&. Other bootloaders will just ignore it if set:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
boot {
        bootdev /dev/sda
}
.ft P
.fi
.UNINDENT
.UNINDENT
.SS \fIboot :: terminal\fP
.sp
Specifies the terminal mode: either \(dqserial\(dq or \(dqvideo\(dq. Defaults to \(dqvideo\(dq.
This setting affects your boot\-loader input/output as well as your kernel\(aqs
output.
.sp
If you set it to \(dqserial\(dq, use the \(dqserial\(dq section (see below) to set serial
options.
.SH SERIAL SECTION
.sp
Specifies the serial port settings for both boot\-loader and kernel. Possible
values are:
.sp
\fIunit\fP (tty number, defaults to 0)
\fIspeed\fP (bps, defaults to 115200)
\fIword\fP (word size, defaults to 8)
\fIparity\fP (yes/no, defaults to no)
\fIstop\fP (stop bit, defaults to 1)
.SH DEFAULT AND BOOT ENTRY SECTIONS
.SS \fIdefault :: type\fP
.sp
Specifies the boot entry type; defaults to \fIlinux\fP\&. Currently, DOS/Windows boot
entries are also supported. Set to one of: \fIlinux\fP, \fIdos\fP, \fImsdos\fP, \fIWindows
2000\fP, \fIwin2000\fP, \fIWindows XP\fP, \fIwinxp\fP, \fIWindows Vista\fP, \fIvista\fP, \fIWindows 7\fP,
\fIwin7\fP, \fIwin8\fP, \fIwin10\fP\&. Here\(aqs how to specify a Windows 7 boot entry, which will automatically
use the proper chainloader +4 parameter to load Microsoft Windows 7:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\(dqMy Windows 7\(dq {
        type win7
        params root=/dev/sda6
}
.ft P
.fi
.UNINDENT
.UNINDENT
.SS \fIdefault :: scan\fP
.sp
This setting specifies one or more directories to scan for kernels and
initrds. Defaults to \fB/boot\fP\&.
.SS \fIdefault :: kernel\fP
.sp
This setting specifies kernel image name, names or patterns, to find kernels to
generate boot menu entries for. The path specified in the \fIscan\fP setting is
searched. Glob patterns are supported, but only one glob pattern may appear per
filename. The special pattern \fI[\-v]\fP is used to match a kernel base name (such
as \fIbzImage\fP) plus all kernels with an optional version suffix beginning with a
\fI\-\fP, such as \fIbzImage\-2.6.24\fP\&. In addition, arbitrary globs can be specified,
such as \fIbzImage[\-2.6.*]\fP If more than one kernel image matches a pattern, or
more than one kernel image is specified, then more than one boot entry will be
created using the settings in this section.
.SS \fIdefault :: initrd\fP
.sp
This setting specifies initrd/initramfs image(s) to load with the menu entry.
If multiple initrds or initramfs images are specified, then \fIall\fP specified
images will be loaded for the boot entry. Linux supports multiple initramfs
images being specified at boot time. Glob patterns are supported. The special
pattern \fI[\-v]\fP is used to find initrd/initramfs images that match the
\fI[\-v]\fP pattern of the current kernel.  For example, if the current menu
entry\(aqs kernel image has a \fI[\-v]\fP pattern of \fI\-2.6.24\fP, then
\fIinitramfs[\-v]\fP will match \fIinitramfs\-2.6.24\fP\&. If the current menu entry
had a \fI[\-v]\fP pattern, but it was blank (in the case of \fIbzImage[\-v]\fP
finding a kernel named \fIbzImage\fP,) then \fIinitramfs[\-v]\fP will match
\fIinitramfs\fP, if it exists.
.SS \fIdefault :: params\fP
.sp
This setting specifies the parameters passed to the kernel. This option
appearing in the \fIdefault\fP section can be extended in specific menu sections
by using the \fI+=\fP operator. The special parameters \fIroot=auto\fP,
\fIrootfstype=auto\fP and \fIreal_root=auto\fP are supported, which will be
replaced with similar settings with the \fIauto\fP string replaced with the
respective setting from \fI/etc/fstab\fP\&. Defaults to \fIroot=auto
rootfstype=auto\fP\&.
.SH DISPLAY SECTION
.SS \fIdisplay :: gfxmode\fP
.sp
Specifies the video mode to be used by the boot loader\(aqs menus. This value is
also inherited and used as the video mode for the kernel when a graphical boot
(\fIuvesafb\fP, \fIvesafb\-tng\fP) is used. This option is only supported for
\fIgrub\fP\&. Default value is \(dqtext\(dq for MBR booting, or \(dq640x480\(dq for UEFI booting.
.SS \fIdisplay :: background\fP
.sp
Specifies the graphical image to display at boot. The specified file should
exist within \fB/boot\fP, and the path to the file should be specified relative

plugins/boot.py
#!/usr/bin/python3

import os
import sys
from ego.module import EgoModule
from ego.output import Output, mesg

import funtoo.boot.helper
import funtoo.boot.extensions

from funtoo.boot.config import BootConfigFile
from funtoo.core.config import ConfigFileError
from funtoo.boot.extension import ExtensionError, BootLoaderMenu


class Module(EgoModule):
	@property
	def boot_config_file(self):
		outpath = os.path.join(self.config.root_path, "etc/boot.conf")
		if self.config.root_path != "/":
			Output.warning("Using boot configuration file %s." % outpath)
		return outpath

	def get_boot_config(self):
		cfgfile = self.boot_config_file
		try:
			config = BootConfigFile(cfgfile)
			if not config.fileExists():
				Output.error("Master configuration file %s does not exist." % cfgfile)
			return config
		except ConfigFileError as e:
			Output.fatal("Error reading %s: %s." % (cfgfile, str(e)))

	def setup(self):

		self.boot_config = self.get_boot_config()

		self._ext = None
		self.ext_name = None
		self.ext_module = None

	def cleanup(self, success, quiet=False):
		""" prints accumulated errors and warnings and exits appropriately """
		print()
		if self._ext is not None:
			# unmounts any non-/boot filesystems we may have mounted for scanning.
			# /boot unmounting is handled via another process.
			if self.config.root_path == "/":
				self._ext.resolver.unmount_if_necessary()
		warnings = False
		for msgtype, msg in self.msgs:
			if msgtype == "warn":
				warnings = True
			if not quiet:
				mesg(msgtype, msg)
		if not success or "error" in map(lambda x: x[0], self.msgs):
			mesg("fatal", "Did not complete successfully.")
			print()
			sys.exit(1)
		else:
			outstr = "Completed successfully"
			if warnings:
				outstr += " with warnings."
			else:
				outstr += "."
			mesg("norm", outstr)
			print()
			sys.exit(0)

	def update(self, boot_menu: BootLoaderMenu, check=False, quiet=False, cleanup=True, boot_options=None):
		"""
		Perform traditional boot-update action of updating boot-loader configuration based on /etc/boot.conf.
		:return:
		"""

		if boot_options is None:
			boot_options = {}

		for invalid in self.boot_config.validate():
			self.msgs.append(["warn", 'invalid config setting "{iv}"; ignored.'.format(iv=invalid)])

		if check is True:
			self.msgs.append(["norm", "Configuration file {cf} checked.".format(cf=self.boot_config_file)])
			self.cleanup(True)

		if os.geteuid() != 0:
			Output.fatal("Updating boot configuration requires root privileges.")

		extension = self.get_extension(boot_options)
		mesg("norm", "Generating config for {gen}...".format(gen=self.ext_name))
		print()

		# Before loading extension, we want to auto-mount boot if it isn't
		# already mounted:

		imountedit = False
		fstabinfo = None

		if self.config.root_path == "/":
			fstabinfo = funtoo.boot.helper.fstabInfo("/")

			if fstabinfo.hasEntry("/boot"):
				if not os.path.ismount("/boot"):
					mesg("debug", "Mounting filesystem /boot...")
					os.system("mount /boot")
					imountedit = True
			else:
				mesg("info", "No /etc/fstab entry for /boot; not mounting.")

		# regenerate config:
		try:
			success = extension.regenerate(boot_menu)
			if success and boot_menu.success and not quiet:
				boot_menu.show()

			# If we mounted /boot, we should unmount it:
			if imountedit:
				mesg("debug", "Unmounting /boot")
				os.system("umount /boot")

			if cleanup:
				self.cleanup(boot_menu.success)
			return boot_menu
		except ExtensionError as e:
			self.msgs.append(["fatal", e])
			if cleanup:
				self.cleanup(False)
				return None

	def get_extension(self, boot_options):
		if self._ext is None:
			success = True
			self.ext_name = self.boot_config["boot/generate"]
			if self.ext_name == "":
				success = False
				self.msgs.append(["fatal", "boot/generate does not specify a valid boot loader to generate a config for."])
			if self.ext_name not in funtoo.boot.extensions.__all__:
				success = False
				self.msgs.append(
					["fatal", 'extension for boot loader "%s" (specified in boot/generate) not found.' % self.ext_name]
				)
			if not success:
				self.cleanup(False)
			# Dynamically import the proper extension module (ie. grub.py,
			# grub-legacy.py, lilo.py):
			extname = "funtoo.boot.extensions.{gen}".format(gen=self.ext_name)
			__import__(extname)
			self.ext_module = sys.modules[extname]
			self._ext = self.ext_module.getExtension(self.boot_config, self.config, boot_options, self)
		return self._ext

	def handle_show_action(self):
		"""Perform boot-update --show action -- show a specific boot.conf configuration setting."""
		print(self.boot_config[self.options.show])

	def handle_show_defaults_action(self):
		"""Perform boot-update --show-defaults action."""
		print("# These are the default settings that can be overridden by")
		print("# the /etc/boot.conf file.")
		print("")
		for line in self.boot_config.parent.dump():
			if not line.startswith("#"):
				sys.stdout.write(line)

	def microcode_action(self):
		from funtoo.boot.resolver import Resolver

		resolver = Resolver(self.boot_config, self.config, self.boot_options, self)
		success = resolver.microcode_regenerate()
		self.cleanup(success=success)

	def set_default_action(self):
		"""Perform the boot-update --set-default action to set a default kernel."""
		if os.geteuid() != 0:
			Output.fatal("Updating the default kernel requires root privileges.")
		# Record default selection in /etc/boot.d...
		found = self.boot_config.idmapper.set_default_kname(sys.argv[2])
		if found:
			# That was successful. Now regenerate our config...
			boot_menu = self.update_action(quiet=True)
			ext = self.get_extension()
			if hasattr(ext, "_set_default"):
				# our extension has a built-in method to set a default boot setting -- use it, to keep it in sync.
				# (Generally, only benefit of doing this is to overwrite an existing setting that may have been set
				# by the user using grub-reboot since last boot-update was run.)
				ext._set_default(boot_menu)
			self.msgs.append(["info", "%s set to default kernel." % sys.argv[2]])
		else:
			self.msgs.append(["error", "Could not find specified kernel image."])
		self.cleanup(success=found)

	def success_action(self):
		if os.geteuid() != 0:
			Output.fatal("This action requires root privileges.")
		# Update our record of the last kernel booted:
		self.boot_config.idmapper.update_last_id()

		# If a kernel is waiting to be promoted to default, then do it:
		promoted, kname = self.boot_config.idmapper.promote_kernel()
		if promoted:
			boot_menu = BootLoaderMenu(self.get_extension(), self.boot_config)
			self.update(boot_menu, quiet=True, cleanup=False)
			if boot_menu.has_kname(kname):
				self.msgs.append(["info", "Boot success -- %s promoted to default kernel." % kname])
			else:
				self.msgs.append(["warn", "Could not find the kernel %s to promote." % kname])
			self.cleanup(boot_menu.success)
		else:
			self.msgs.append(["warn", "Unable to find a kernel to promote."])
			self.cleanup(False)

	def attempt_action(self):
		if self.options.identifier != "default":
			boot_menu = BootLoaderMenu(
				self.get_extension(self.boot_options), self.boot_config, user_specified_attempt_identifier=self.options.identifier
			)
			self.update(boot_menu, quiet=False, cleanup=True)
		else:
			self.boot_config.idmapper.remove_promote_setting()
			self.msgs.append(["info", "Any attempted kernel setting has been wiped -- default will be used."])
			return self.update_action()

	@property
	def boot_options(self):
		return {"device-shift": self.options.device_shift}

	def update_action(self, check=False, quiet=False) -> BootLoaderMenu:
		boot_menu = BootLoaderMenu(self.get_extension(self.boot_options), self.boot_config)
		self.update(boot_menu, quiet=quiet, cleanup=True, check=check, boot_options=self.boot_options)
		return boot_menu

	def show_action(self):
		if "defaults" in self.options.sub_action:
			self.handle_show_defaults_action()
		else:
			print(self.options.sub_action)
			print("ARGH!")

	def add_arguments(self, parser):
		parser.add_argument(
			"--show-defaults", "--showdefaults", action="store_true", help="Show default settings for /etc/boot.conf."
		)
		parser.add_argument("--show", default=None, metavar="sect/val", help="Echo a specific configuration setting.")
		parser.add_argument(
			"--set-default",
			default=None,
			metavar="path-to-kernel-img",
			help="Set a default kernel image to boot using /etc/boot.d.",
		)
		parser.add_argument(
			"--check", action="store_true", help="Check the validity of the %s file." % self.boot_config_file
		)
		parser.add_argument(
			"--device-shift",
			default=None,
			metavar="sda,sdb",
			help="Modify disk references FROM,TO -- for generating configs for temp. mounted disks.",
		)
		subparsers = parser.add_subparsers(title="actions", dest="action")
		update_parser = subparsers.add_parser("update", help="Update boot loader configuration based on /etc/boot.conf.")
		update_parser.set_defaults(handler=self.update_action)
		success_parser = subparsers.add_parser("success", help="Record a successful boot.")
		success_parser.set_defaults(handler=self.success_action)
		attempt_parser = subparsers.add_parser(
			"attempt", help="Attempt to boot a new kernel without actually changing default."
		)
		attempt_parser.set_defaults(handler=self.attempt_action)
		attempt_parser.add_argument("identifier")
		microcode_parser = subparsers.add_parser(
			"microcode", help="Regenerate microcode without actually (re)generating boot loader files.."
		)
		microcode_parser.set_defaults(handler=self.microcode_action)

	def handle(self):
		handler = getattr(self.options, "handler", None)
		if handler is not None:
			# an action was specified:
			handler()
		else:
			# no action -- just an option. Call the right one:
			if self.options.show_defaults:
				self.handle_show_defaults_action()
			elif self.options.show is not None:
				self.handle_show_action()
			elif self.options.set_default is not None:
				self.set_default_action()
			elif self.options.check:
				self.update_action(check=True)
			else:
				self.update_action()


# vim: ts=4 sw=4 noet
